File: backend/skills/ast_audit.go
package skills

import (
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
)

type ASTFinding struct {
	File        string `json:"file"`
	Line        int    `json:"line"`
	PatternID   string `json:"pattern_id"`
	Description string `json:"description"`
}

var (
	rxDynamicImport   = regexp.MustCompile(`import_module\s*\(`)
	rxComputedImport  = regexp.MustCompile(`__import__\s*\(\s*[^'"]`)
	rxDynamicGetattr  = regexp.MustCompile(`getattr\s*\(\s*[^,]+,\s*[^'"]`)
	rxDictAccess      = regexp.MustCompile(`__dict__\s*\[\s*[^'"]`)
	rxImportlibImport = regexp.MustCompile(`(?:^|\n)\s*(?:import\s+importlib|from\s+importlib\s+import)`)
)

func scanSource(content, relPath string) []ASTFinding {
	var findings []ASTFinding
	lines := strings.Split(content, "\n")

	for i, line := range lines {
		lineNum := i + 1
		trimmed := strings.TrimSpace(line)
		if trimmed == "" || strings.HasPrefix(trimmed, "#") {
			continue
		}

		if rxDynamicImport.MatchString(line) {
			findings = append(findings, ASTFinding{
				File:        relPath,
				Line:        lineNum,
				PatternID:   "dynamic_import",
				Description: "importlib.import_module() — loads arbitrary modules at runtime",
			})
		}
		if rxComputedImport.MatchString(line) {
			findings = append(findings, ASTFinding{
				File:        relPath,
				Line:        lineNum,
				PatternID:   "dynamic_import_computed",
				Description: "__import__ with non-literal module name",
			})
		}
		if rxDynamicGetattr.MatchString(line) {
			findings = append(findings, ASTFinding{
				File:        relPath,
				Line:        lineNum,
				PatternID:   "dynamic_getattr",
				Description: "getattr with non-literal attribute name",
			})
		}
		if rxDictAccess.MatchString(line) {
			findings = append(findings, ASTFinding{
				File:        relPath,
				Line:        lineNum,
				PatternID:   "dict_access",
				Description: "__dict__[<computed>] — dynamic attribute access",
			})
		}
		if rxImportlibImport.MatchString(line) {
			findings = append(findings, ASTFinding{
				File:        relPath,
				Line:        lineNum,
				PatternID:   "importlib_import",
				Description: "import importlib — enables dynamic module loading",
			})
		}
	}

	return findings
}

func ASTScanPath(path string) ([]ASTFinding, error) {
	fi, err := os.Stat(path)
	if err != nil {
		return nil, err
	}

	if !fi.IsDir() {
		if strings.ToLower(filepath.Ext(path)) != ".py" {
			return nil, nil
		}
		data, err := os.ReadFile(path)
		if err != nil {
			return nil, err
		}
		return scanSource(string(data), filepath.Base(path)), nil
	}

	var findings []ASTFinding
	ignoredDirs := map[string]bool{
		"__pycache__":   true,
		".venv":         true,
		"venv":          true,
		"node_modules":  true,
		".git":          true,
	}

	err = filepath.Walk(path, func(p string, info os.FileInfo, err error) error {
		if err != nil {
			return nil
		}
		if info.IsDir() {
			if ignoredDirs[info.Name()] {
				return filepath.SkipDir
			}
			return nil
		}
		if strings.ToLower(filepath.Ext(p)) != ".py" {
			return nil
		}
		rel, err := filepath.Rel(path, p)
		if err != nil {
			rel = filepath.Base(p)
		}
		data, err := os.ReadFile(p)
		if err == nil {
			findings = append(findings, scanSource(string(data), filepath.ToSlash(rel))...)
		}
		return nil
	})

	return findings, err
}

func FormatASTReport(findings []ASTFinding, skillName string) string {
	header := "AST deep scan"
	if skillName != "" {
		header = fmt.Sprintf("AST deep scan: %s", skillName)
	}

	if len(findings) == 0 {
		return fmt.Sprintf("%s\n  No dynamic import/access patterns detected.", header)
	}

	sort.Slice(findings, func(i, j int) bool {
		if findings[i].File != findings[j].File {
			return findings[i].File < findings[j].File
		}
		return findings[i].Line < findings[j].Line
	})

	var lines []string
	lines = append(lines, header, fmt.Sprintf("  %d finding(s):", len(findings)))

	currentFile := ""
	for _, f := range findings {
		if f.File != currentFile {
			currentFile = f.File
			lines = append(lines, fmt.Sprintf("  %s", f.File))
		}
		lines = append(lines, fmt.Sprintf("    L%d  %s  — %s", f.Line, f.PatternID, f.Description))
	}
	lines = append(lines, "", "  Note: diagnostic hints for human review, not security verdicts.")

	return strings.Join(lines, "\n")
}

File: backend/skills/ast_audit.ts
// PORT: backend/skills/ast_audit.go

import { Effect } from "effect";
import fs from "node:fs";
import path from "node:path";

export interface ASTFinding {
  File: string;
  Line: number;
  PatternID: string;
  Description: string;
}

const rxDynamicImport = /import_module\s*\(/;
const rxComputedImport = /__import__\s*\(\s*[^'"]/;
const rxDynamicGetattr = /getattr\s*\(\s*[^,]+,\s*[^'"]/;
const rxDictAccess = /__dict__\s*\[\s*[^'"]/;
const rxImportlibImport = /(?:^|\n)\s*(?:import\s+importlib|from\s+importlib\s+import)/;

function scanSource(content: string, relPath: string): ASTFinding[] {
  const findings: ASTFinding[] = [];
  const lines = content.split("\n");

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];
    const lineNum = i + 1;
    const trimmed = line.trim();
    if (trimmed === "" || trimmed.startsWith("#")) {
      continue;
    }

    if (rxDynamicImport.test(line)) {
      findings.push({
        File: relPath,
        Line: lineNum,
        PatternID: "dynamic_import",
        Description: "importlib.import_module() — loads arbitrary modules at runtime",
      });
    }
    if (rxComputedImport.test(line)) {
      findings.push({
        File: relPath,
        Line: lineNum,
        PatternID: "dynamic_import_computed",
        Description: "__import__ with non-literal module name",
      });
    }
    if (rxDynamicGetattr.test(line)) {
      findings.push({
        File: relPath,
        Line: lineNum,
        PatternID: "dynamic_getattr",
        Description: "getattr with non-literal attribute name",
      });
    }
    if (rxDictAccess.test(line)) {
      findings.push({
        File: relPath,
        Line: lineNum,
        PatternID: "dict_access",
        Description: "__dict__[<computed>] — dynamic attribute access",
      });
    }
    if (rxImportlibImport.test(line)) {
      findings.push({
        File: relPath,
        Line: lineNum,
        PatternID: "importlib_import",
        Description: "import importlib — enables dynamic module loading",
      });
    }
  }

  return findings;
}

export function ASTScanPath(filePath: string): Effect.Effect<ASTFinding[], Error> {
  return Effect.try({
    try: () => {
      const fi = fs.statSync(filePath);
      if (!fi.isDirectory()) {
        if (path.extname(filePath).toLowerCase() !== ".py") {
          return [];
        }
        const data = fs.readFileSync(filePath, "utf8");
        return scanSource(data, path.basename(filePath));
      }

      const findings: ASTFinding[] = [];
      const ignoredDirs = new Set(["__pycache__", ".venv", "venv", "node_modules", ".git"]);

      const walk = (dir: string) => {
        let entries: string[];
        try {
          entries = fs.readdirSync(dir);
        } catch {
          return;
        }

        for (const entry of entries) {
          const full = path.join(dir, entry);
          let stat: fs.Stats;
          try {
            stat = fs.statSync(full);
          } catch {
            continue;
          }

          if (stat.isDirectory()) {
            if (ignoredDirs.has(entry)) {
              continue;
            }
            walk(full);
          } else {
            if (path.extname(full).toLowerCase() !== ".py") {
              continue;
            }
            let rel = path.relative(filePath, full);
            if (!rel) {
              rel = path.basename(full);
            }
            // Normalize separator to slash like Go's filepath.ToSlash
            const relSlash = rel.split(path.sep).join("/");
            try {
              const data = fs.readFileSync(full, "utf8");
              findings.push(...scanSource(data, relSlash));
            } catch {}
          }
        }
      };

      walk(filePath);
      return findings;
    },
    catch: (cause) => (cause instanceof Error ? cause : new Error(String(cause))),
  });
}

export function FormatASTReport(findings: ASTFinding[], skillName: string): string {
  let header = "AST deep scan";
  if (skillName !== "") {
    header = `AST deep scan: ${skillName}`;
  }

  if (findings.length === 0) {
    return `${header}\n  No dynamic import/access patterns detected.`;
  }

  const sorted = [...findings].sort((a, b) => {
    if (a.File !== b.File) {
      return a.File.localeCompare(b.File);
    }
    return a.Line - b.Line;
  });

  const lines: string[] = [];
  lines.push(header, `  ${sorted.length} finding(s):`);

  let currentFile = "";
  for (const f of sorted) {
    if (f.File !== currentFile) {
      currentFile = f.File;
      lines.push(`  ${f.File}`);
    }
    lines.push(`    L${f.Line}  ${f.PatternID}  — ${f.Description}`);
  }
  lines.push("", "  Note: diagnostic hints for human review, not security verdicts.");

  return lines.join("\n");
}

/*
PORT STATUS
source path: backend/skills/ast_audit.go
source lines: 168
draft lines: 167
confidence: high
status: phase_b_compile
*/
